executor-utils/src/expression_evaluator.rs

```diff
@@ -70,35 +78,82 @@ impl<F: FieldElement> TraceValues<F> for RowTraceValues<'_, F> {
     }
 }
 
+#[derive(Default)]
+pub struct OwnedGlobalValues<T> {
+    pub public_values: BTreeMap<String, T>,
+    pub challenge_values: BTreeMap<u64, T>,
+}
+
+impl<T: Clone> GlobalValues<T> for &OwnedGlobalValues<T> {
+    fn get_public(&self, public: &str) -> T {
+        self.public_values[public].clone()
+    }
+
+    fn get_challenge(&self, challenge: &Challenge) -> T {
+        self.challenge_values[&challenge.id].clone()
+    }
+}
+
 /// Evaluates an algebraic expression to a value.
-pub struct ExpressionEvaluator<'a, T, SV> {
-    trace_values: SV,
+pub struct ExpressionEvaluator<'a, T, Expr, TV, GV> {
+    trace_values: TV,
+    global_values: GV,
     intermediate_definitions: &'a BTreeMap<AlgebraicReferenceThin, Expression<T>>,
-    challenges: &'a BTreeMap<u64, T>,
     /// Maps intermediate reference to their evaluation. Updated throughout the lifetime of the
     /// ExpressionEvaluator.
-    intermediates_cache: BTreeMap<AlgebraicReferenceThin, T>,
+    intermediates_cache: BTreeMap<AlgebraicReferenceThin, Expr>,
+    to_expr: fn(&T) -> Expr,
+    one_expr: fn() -> Expr,
```

Comment:
I wouldn't need this if I could do `Expr: From<T> + One`. But if these are backend types, I can't implement a foreign trait on them...

Reply:
One solution would be using the new type pattern in the backends. If it removes this weirdness, it could be worth it.

Reply:
Not sure what you mean, could you elaborate? Do you think this is something that should be done in this PR?

Reply:
What I understood from the code: say in stwo, `Expr` is a type `StwoExpr` from the stwo crate. So we can't implement traits like `One` and `<T: FieldElement> From<T>` on it. By wrapping `StwoExpr` in a new type `StwoExprWrapper`, we can.

Reply:
Yes, I think that would be pretty annoying though, because we'd have a lot of boilerplate code to also implement traits like `Add`, `Sub`, `Mul`, ...

Reply:
Good point.. Can we at least remove `one_expr` and use `to_expr(&T::ONE)` where needed?

Reply:
Right, that worked, nice!

Reply:
Annoying, even reducing the problem to the backend implementing `From<Scalar>` (instead of `From<T>`) is not enough because halo2 doesn't implement anything on `Expr`.